cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1128 @@
+
+#include <cub/detail/uninitialized_copy.cuh>
+#include <cub/device/device_for.cuh>
+#include <cub/util_arch.cuh>
+#include <cub/util_type.cuh>
+
+#include <thrust/detail/raw_reference_cast.h>
+#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
+#include <thrust/type_traits/is_contiguous_iterator.h>
+#include <thrust/type_traits/is_trivially_relocatable.h>
+
+#include <cuda/cmath>
+#include <cuda/ptx>
+#include <cuda/std/__algorithm/clamp.h>
+#include <cuda/std/__algorithm/max.h>
+#include <cuda/std/__algorithm/min.h>
+#include <cuda/std/array>
+#include <cuda/std/expected>
+#include <cuda/std/tuple>
+#include <cuda/std/type_traits>
+#include <cuda/std/utility>
+
+#include <cooperative_groups.h>
+#include <cooperative_groups/memcpy_async.h>
+
+CUB_NAMESPACE_BEGIN
+
+// the ublkcp kernel needs PTX features that are only available and understood by CTK 12 and later
+#if _CCCL_CUDACC_VER_MAJOR >= 12
+#  define _CUB_HAS_TRANSFORM_UBLKCP
+#endif // _CCCL_CUDACC_VER_MAJOR >= 12
+
+namespace detail
+{
+namespace transform
+{
+#ifdef __cpp_fold_expressions // C++17
+template <typename... Its>
+constexpr auto loaded_bytes_per_iteration() -> std::size_t
+{
+  return (sizeof(value_t<Its>) + ... + 0);
+}
+#else
+constexpr std::size_t sum()
+{
+  return 0;
+}
+
+template <typename... Ts>
+constexpr std::size_t sum(std::size_t head, Ts... tail)
+{
+  return head + sum(tail...);
+}
+
+template <typename... Its>
+constexpr auto loaded_bytes_per_iteration() -> std::size_t
+{
+  return sum(sizeof(value_t<Its>)...);
+}
+#endif
+
+enum class Algorithm
+{
+  fallback_for,
+  prefetch,
+  unrolled_staged,
+  memcpy_async
+#ifdef _CUB_HAS_TRANSFORM_UBLKCP
+  ,
+  ublkcp
+#endif // _CUB_HAS_TRANSFORM_UBLKCP
+};
+
+// TODO(bgruber): only needed so we can instantiate the kernel generically for any policy. Remove when fallback_for is
+// dropped.
+template <typename, typename Offset, typename F, typename RandomAccessIteartorOut, typename... RandomAccessIteartorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::fallback_for>,
+  Offset,
+  int,
+  F,
+  RandomAccessIteartorOut,
+  RandomAccessIteartorIn...)
+{}
+
+template <int BlockThreads>
+struct prefetch_policy_t
+{
+  static constexpr int BLOCK_THREADS = BlockThreads;
+  // items per tile are determined at runtime. these (inclusive) bounds allow overriding that value via a tuning policy
+  static constexpr int ITEMS_PER_THREAD_NO_INPUT = 2; // items per thread when no input streams exits (just filling)
+  static constexpr int MIN_ITEMS_PER_THREAD      = 1;
+  static constexpr int MAX_ITEMS_PER_THREAD      = 32;
+};
+
+// Prefetches (at least on Hopper) a 128 byte cache line. Prefetching out-of-bounds addresses has no side effects
+template <typename T>
+_CCCL_DEVICE void prefetch(const T* addr)
+{
+  // TODO(bgruber): prefetch to L1 may be even better
+  asm volatile("prefetch.L2 [%0];" : : "l"(addr) : "memory");
```

Comment:
`prefetch.global.L2` could help the compiler instead of generic memory addressing. We could even verify this property with `assert(__isGlobal(addr))`

Reply:
Fine for me.